### PR TITLE
Demangler: fix demangling and remangling of associated type paths.

### PR DESCRIPTION
**Pull request by eeckstein**
fixes SR-6138

### CODE REVIEW

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

File: test/IRGen/assoctypepath.swift [hunk 1]
```diff
@@ -0,0 +1,22 @@
+// RUN: %target-swift-frontend -primary-file %s -emit-ir | %FileCheck %s
```
Comment (slavapestov) [inline test/IRGen/assoctypepath.swift]: This will fail because there are no CHECK: directives -- can you add some for the mangled symbols?

Comment (eeckstein) [reply test/IRGen/assoctypepath.swift]: oh, thanks! Yeah, that's what I wanted but apparently not did

**Comment by eeckstein** [comment]
@swift-ci  smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Comment by eeckstein** [comment]
@swift-ci smoke test and merge

**Pull request merged**